### PR TITLE
Feature/transition solver API

### LINKED ISSUE #581: Refactor reduction strategy code to enable JIT of entire coreset construction
### What's the new feature?

Related to issues #534 & #544 - update the implementations of reduction strategies to allow the entire codebase to be JIT'ed using Equinox, which enables fast computation of several coresets. 

Clean up test_coresubset.py to not have to define a metric_x 

### What value does this add?

Improved integration with standard JAX approaches & removes side-effects that interfere with expected JIT performance.

### Is there an alternative you've considered?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tc85324**
### PR Type
Closes #581 #265 #102

- Feature
- Refactoring
- Documentation content changes
- Tests

### Description
Transitions the codebase over to the new solver API.

Removes the performance tests in favour of JIT compilation success checks in `test_kernels.py` and `test_solver.py`. Detailed rationale can be found in the commit messages. Basic logic is to defer the "performance" checks to benchmarking and the JIT-ability to testing (this markedly simplifies the removal process as the performance tests don't require changing).

### How Has This Been Tested?
All tests pass.

### Does this PR introduce a breaking change?
Removal of refine:
- The solvers now provide access to the refinement procedures. The random and reverse variants can be obtained by randomly partitioning the coresubsets and applying refinement to these reduced random sets and reversing the data respectively.

Removal of reduction:
- All functionality provided by reduction is now handled by the solvers.
- ~~Except `MapReduce` which has been removed due to its JIT incompatible recursive design. (This/similar approaches can be re-introduced in a future pull)~~ EDIT: a simple solution was found here by using padding and recursing based on a traceable value (an array shape).

Removal of couplings to coreset data-structures
- coreset.compute_metric: retained
- coreset.solve_weights: retained
- coreset.render: no replacement planned
- coreset.format: no replacement planned
- coreset.copy_fit_shallow: Redundant due to use of frozen dataclasses
- coreset.validate_fitted: a coreset datastructure is always considered as 'fitted'

Removal of DataReaders in favour of the new Data classes
- ArrayData and DataReader classes are gone

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by db091756** [comment]
Changelog needs updating

File: coreax/__init__.py [hunk 1]
```diff
@@ -59,8 +59,7 @@
     NystromApproximateKernel,
```
Comment (db091756) [inline coreax/__init__.py:59]: The __init__ doc is out of date talking about "See :class:`@coreax.kernel.Kernel`" for flattening methods.

File: coreax/data.py [hunk 2]
```diff
@@ -247,16 +78,6 @@ def normalize(self, *, preserve_zeros: bool = False) -> Data:
...
-def as_data(x: Any) -> Data:
-    """Cast 'x' to a data instance."""
-    return x if isinstance(x, Data) else Data(x)
-
-
-def is_data(x: Any) -> bool:
-    """Return boolean indicating if 'x' is an instance of 'coreax.data.Data'."""
-    return isinstance(x, Data)
-
-
```
Comment (db091756) [inline coreax/data.py:259]: missed doc issue "defining the responses of the supervised paired " -> "defining the responses of the supervised dataset paired "

File: coreax/__init__.py (hunk 1 above)
Comment (tc85324) [reply coreax/__init__.py:59]: Are you happy for me to simply remove everything after the first paragraph, seeing as everything else is now pretty much redudant?

File: tests/unit/test_coresubset.py
Comment (pc532627) [inline tests/unit/test_coresubset.py:1]: Just as a sense check, 1,252 lines have been removed here. Are you aware of any functionality that was previously tested that is no longer tested?

File: examples/herding_basic.py [hunk 3]
```diff
@@ -84,57 +80,47 @@ def main(out_path: Path | None = None) -> tuple[float, float, float]:
...
-    )
-    herding_object = KernelHerding(
-        herding_key, kernel=SquaredExponentialKernel(length_scale=length_scale)
-    )
-    herding_object.fit(
-        original_data=data, strategy=SizeReduce(coreset_size=coreset_size)
-    )
+    sample_key, rpc_key, stein_key = random.split(random.key(random_seed), num=3)
+    herding_solver = KernelHerding(coreset_size, kernel)
+    herding_coreset, _ = eqx.filter_jit(herding_solver.reduce)(data)
```
Comment (db091756) [inline examples/herding_basic.py:97]: What is the reasoning behind needing ` eqx.filter_jit` wrapping the reduce calls in the examples?

File: coreax/__init__.py (hunk 1 above)
Comment (db091756) [reply coreax/__init__.py:59]: Yeah I think that is reasonable

File: coreax/data.py (hunk 2 above)
Comment (db091756) [reply coreax/data.py:259]: Yep happy

File: examples/herding_basic.py (hunk 3 above)
Comment (db091756) [reply examples/herding_basic.py:97]: If this is necessary should this be included in documentation for the solvers?

File: tests/unit/conftest.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+"""Fixtures used across multiple unit tests modules."""
+
+from collections.abc import Callable
+
+import equinox as eqx
+import pytest
+
+
+@pytest.fixture(params=["with_jit", "without_jit"], scope="class")
+def jit_variant(request) -> Callable[[Callable], Callable]:
```
Comment (db091756) [inline tests/unit/conftest.py:24]: request needs type hint

File: examples/herding_basic.py (hunk 3 above)
Comment (tc85324) [reply examples/herding_basic.py:97]: Demonstrating that you can JIT compile the reduction step. We don't want to automatically apply JIT decoration to the solver reduce methods for a few reasons. Firstly, it leads to a lot of repeat decoration across the codebase. Secondly, and more importantly, we want to defer JIT decoration to the largest possible scope which enables the compiler to perform more/better optimisations.

We use `eqx.filter_jit` instead of `jax.jit` because it provides a more convenient interface (automatically selects jax and numpy arrays as traceable and fixes everything else as static) [see here](https://docs.kidger.site/equinox/api/transformations/#equinox.filter_jit).

**Review by db091756 — COMMENTED** [review]
A few small changes but overall happy

File: examples/herding_basic.py (hunk 3 above)
Comment (db091756) [reply examples/herding_basic.py:97]: Makes sense!

File: tests/unit/test_data.py
Comment (pc532627) [inline tests/unit/test_data.py:1]: Just as a sense check, some test files, e.g. `test_networks.py` have this at the bottom:

```
if __name__ == "__main__":
    unittest.main()
```

but several others do not. I suspect it's the choice of pytest vs unittest - in the style guide we permit either so is there a reason not to include the above code snippet in all the code files?

**Review by pc532627 — CHANGES_REQUESTED** [review]
Some very minor changes/questions but looks good - I'm going to check the documentation now so may follow up with minor comments

File: examples/david_map_reduce_weighted.py [hunk 5]
```diff
@@ -158,52 +155,42 @@ def main(
...
 
     print("Choosing random subset...")
     # Generate a coreset via uniform random sampling for comparison
-    random_sample_object = RandomSample(sample_key, unique=True)
-    random_sample_object.fit(
-        original_data=data,
-        strategy=SizeReduce(coreset_size=coreset_size),
-    )
+    random_solver = RandomSample(coreset_size, sample_key, unique=True)
+    random_coreset, _ = eqx.filter_jit(random_solver.reduce)(data)
```
[the remark this answers is not in the document]
Comment (tc85324) [reply examples/david_map_reduce_weighted.py:173]: As above comment; see here https://github.com/gchq/coreax/pull/632#discussion_r1620424830

File: tests/unit/test_data.py
Comment (tc85324) [reply tests/unit/test_data.py:1]: Whenever unittest is completely unsued I've been removing these. Because we are using pytest for running, I'm not 100% sure if we ever need these, but I'm honestly not sure.

File: examples/david_map_reduce_weighted.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pc532627) [reply examples/david_map_reduce_weighted.py:173]: It might be worth documenting this somewhere, although the examples could be documentation enough themselves. What do you think for a user trying to use the library for the first time?

[the remark this answers is not in the document]
Comment (tc85324) [reply examples/david_map_reduce_weighted.py:173]: I'm happy to explain why we are doing this in the `quickstart.rst` part of the docs

[the remark this answers is not in the document]
Comment (pc532627) [reply examples/david_map_reduce_weighted.py:173]: @tc85324 I'd be happy with that, thanks

**Review by pc532627 — APPROVED** [review]
